packages/tokamak/contracts-bedrock/test/L1CrossDomainMessenger.t.sol

```diff
@@ -199,58 +201,64 @@ contract L1CrossDomainMessenger_Test is Messenger_Initializer {
-
-    //     vm.expectEmit(true, true, true, true);
-
-    //     emit RelayedMessage(hash);
-
-    //     vm.etch(target, address(0).code);
-    //     vm.prank(address(sender));
-    //     L1Messenger.relayMessage(
-    //         Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }), // nonce
-    //         sender,
-    //         target,
-    //         value,
-    //         0,
-    //         hex"1111"
-    //     );
-
-    //     assertEq(address(L1Messenger).balance, 0);
-    //     assertEq(address(target).balance, value);
-    //     assertEq(L1Messenger.successfulMessages(hash), true);
-    //     assertEq(L1Messenger.failedMessages(hash), true);
-    // }
+    /// @dev Tests that relayMessage should successfully call the target contract after
+    ///      the first message fails and TON is stuck, but the second message succeeds
+    ///      with a version 1 message.
+    function test_relayMessage_retryAfterFailure_succeeds() external {
+        address target = address(0xabcd);
+        address sender = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
+        uint256 value = 100;
+
+        dealL2NativeToken(address(op), 2 * value);
+        // Approve the L1Messenger contract
+        vm.prank(address(op));
+        IERC20(l2NativeToken).approve(address(L1Messenger), type(uint256).max);
+
+        vm.expectCall(target, hex"1111");
+
+        bytes32 hash = Hashing.hashCrossDomainMessage(
+            Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }), sender, target, value, 0, hex"1111"
+        );
+
+        vm.store(address(op), bytes32(senderSlotIndex), bytes32(abi.encode(sender)));
+        vm.etch(target, address(new Reverter()).code);
+        vm.prank(address(op));
+        L1Messenger.relayMessage(
+            Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }), // nonce
+            sender,
+            target,
+            value,
+            0,
+            hex"1111"
+        );
+
+        assertEq(address(L1Messenger).balance, 0);
+        assertEq(address(target).balance, 0);
+        assertEq(L1Messenger.successfulMessages(hash), false);
+        assertEq(L1Messenger.failedMessages(hash), true);
+
+        vm.expectEmit(true, true, true, true);
+
+        emit RelayedMessage(hash);
+
+        vm.etch(target, address(0).code);
+        vm.prank(address(sender));
+        L1Messenger.relayMessage(
+            Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }), // nonce
+            sender,
+            target,
+            value,
+            0,
+            hex"1111"
+        );
+
+        assertEq(ERC20Detailed(l2NativeToken).balanceOf(address(L1Messenger)), value);
+        assertEq(ERC20Detailed(l2NativeToken).balanceOf(address(target)), 0);
+        assertEq(address(L1Messenger).balance, 0);
+        assertEq(address(target).balance, 0);
+        assertEq(L1Messenger.successfulMessages(hash), true);
+        assertEq(L1Messenger.failedMessages(hash), true);
```

Comment:
It's great!
How about adding transferFrom(L1Messenger, target, value) (L1Messenger -> value)?
I think it is a useful test!

Reply:
Oh, I'm sorry. I think it is great adding it in `test_relayMessage_succeeds` not in here. 🙇🏼‍♂️

Reply:
@rlgns98kr I've just kept the code from the previous version and fixed it to work with the native token

Reply:
Okay, I understood. I think we can keep your code. I'm sorry for my short thought.

Reply:
Can we resolve conflicts?

Reply:
Yes, a lot of changes, let me resolve this

Reply:
@rlgns98kr resolved

@nguyenzung I've just created a new contract `L2NativeToken_Initializer` and used `L2NativeToken.sol` instead of your contract in `packages/tokamak/contracts-bedrock/test/CommonTest.t.sol`, can you review my PR?